### PR TITLE
Throttle updates for entity-filter

### LINKED ISSUE #3525: Filter card sets config on each hass change
Home Assistant 0.98

The filter card will call setConfig on each `hass` change. This causes a lot of events. It should only call `setConfig` if the entities have changed compared to the previous time it was set.

### PR DESCRIPTION
**Pull request by iantrich**
Fixes https://github.com/home-assistant/home-assistant-polymer/issues/3525

### CODE REVIEW

**Review by balloob — APPROVED** [review]
This is good, and a big improvement and is ok to merge…

But I think that we can do better. 

I think that we should compute the old and the new `entitiesList` value and see if they are different. If they are not, we don't update the config.

**Comment by iantrich** [comment]
Don't update the config unless the entities list changes? I don't get it, there are other things to update in the config

**Comment by bramkragten** [comment]
Here we set the config https://github.com/home-assistant/home-assistant-polymer/blob/ad4df2ea0d16f3e490e7c028a376365f85167826/src/panels/lovelace/cards/hui-entity-filter-card.ts#L73, every time a state changes after your update. But would be better to add a second check and only set the config is the `entitiesList` changed, so after filtering. https://github.com/home-assistant/home-assistant-polymer/blob/ad4df2ea0d16f3e490e7c028a376365f85167826/src/panels/lovelace/cards/hui-entity-filter-card.ts#L56

**Comment by balloob** [comment]
```ts
const isSame = newEntitiesList.length === oldEntitiesList.length && newEntitiesList.every((entity, idx) => entity === oldEntitiesList[idx])
```

**Comment by iantrich** [comment]
Ah, the config of the underlying elements. Got it, thanks for the clarification, guys.

**Comment by iantrich** [comment]
I think I'm missing something as the element's config is still being set when I expect it not to be

**Comment by bramkragten** [comment]
You don't want to save the old config and recalculate everything, just do this:

```js
      private _oldEntities: EntityConfig[] = [];
      const isSame =
        entitiesList.length === this._oldEntities.length &&
        entitiesList.every((entity, idx) => entity === this._oldEntities[idx]);

      if (!isSame) {
        this._oldEntities = entitiesList;
        element.setConfig({ ...this._baseCardConfig!, entities: entitiesList });
      }
```

**Comment by iantrich** [comment]
Thanks @bramkragten, that's much cleaner

File: src/panels/lovelace/cards/hui-entity-filter-card.ts [hunk 1]
```diff
@@ -12,6 +12,8 @@ class EntityFilterCard extends HTMLElement implements LovelaceCard {
   private _config?: EntityFilterCardConfig;
   private _configEntities?: EntityConfig[];
   private _baseCardConfig?: LovelaceCardConfig;
+  private _hass?: HomeAssistant;
+  private _oldEntities: EntityConfig[] = [];
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-entity-filter-card.ts]: The config would never be set when the `entitiesList` will be empty, so let's make it undefined at the start and check for that.
```suggestion
  private _oldEntities?: EntityConfig[];
```

File: src/panels/lovelace/cards/hui-entity-filter-card.ts [hunk 2]
```diff
@@ -62,7 +71,15 @@ class EntityFilterCard extends HTMLElement implements LovelaceCard {
     }
 
     if (element.tagName !== "HUI-ERROR-CARD") {
-      element.setConfig({ ...this._baseCardConfig!, entities: entitiesList });
+      const isSame =
+        entitiesList.length === this._oldEntities.length &&
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-entity-filter-card.ts]:
```suggestion
        entitiesList.length === this._oldEntities && this._oldEntities.length &&
```

**Pull request merged**